listenbrainz/mbid_mapping/reports/tracks_of_the_year.py

```diff
@@ -0,0 +1,122 @@
+            curs.execute("""CREATE INDEX tracks_of_the_year_ndx_user_name
+                                      ON mapping.tracks_of_the_year (user_name)""")
+            mb_conn.commit()
+    except (psycopg2.errors.OperationalError, psycopg2.errors.UndefinedTable) as err:
+        log("mbid mapping: failed to create tracks of the year indexes", err)
+        mb_conn.rollback()
+        raise
+
+
+def fetch_user_list(mb_conn, year):
+    """ Fetch the distinct list of users from tracks listened table """
+
+    with mb_conn.cursor(cursor_factory=psycopg2.extras.DictCursor) as mb_curs:
+        query = """SELECT DISTINCT user_name
+                     FROM mapping.tracks_of_the_year"""
+        mb_curs.execute(query)
+        rows = mb_curs.fetchall()
+
+    return [r[0] for r in rows]
+
+
+def chunks(lst, n):
+    """ Helper function to break a list into chunks """
+    for i in range(0, len(lst), n):
+        yield lst[i:i + n]
+
+
+def fetch_tracks_listened_to(lb_conn, mb_conn, ts):
+    """ Actually fetch the top discoveries for the given year and set of users """
+
+    with lb_conn.cursor(cursor_factory=psycopg2.extras.DictCursor) as lb_curs:
+        with mb_conn.cursor(cursor_factory=psycopg2.extras.DictCursor) as mb_curs:
+            log("create tracks listened table")
+            create_table(mb_conn)
+
+            log("fetch tracks listened to")
+            query = """SELECT user_name
+                            , m.recording_mbid
+                            , md.recording_name
+                            , md.artist_credit_name
+                            , md.artist_mbids
+                            , count(*) AS listen_count
+                         FROM listen l
+                         JOIN mbid_mapping m
+                           ON data->'track_metadata'->'additional_info'->>'recording_msid' = m.recording_msid::TEXT
+                         JOIN mbid_mapping_metadata md
+                           ON m.recording_mbid = md.recording_mbid
+                        WHERE listened_at >= %s
+                          AND listened_at < %s
+                          AND m.recording_mbid is not null
+                     GROUP BY m.recording_mbid, md.recording_name,
+                              md.artist_credit_name, md.artist_mbids, user_name""" % (start_ts, end_ts)
+
+            to_insert = []
+            lb_curs.execute(query)
+            while True:
+                row = lb_curs.fetchone()
+                if not row:
+                    break
+
+                to_insert.append(row)
+                if len(to_insert) >= BATCH_SIZE:
+                    insert_rows(mb_curs, "mapping.tracks_of_the_year", to_insert)
+                    to_insert = []
+                    mb_conn.commit()
+
+            insert_rows(mb_curs, "mapping.tracks_of_the_year", to_insert)
+            mb_conn.commit()
+
+
+def calculate_tracks_of_the_year(year):
+    """
+        Main entry point for creating top discoveries table.
```

Comment:
Let's also add the text from the PR to the docstrings of the respective reports.

Reply:
Good idea.